builder/upcloud/builder.go

```diff
@@ -20,59 +22,64 @@ type Builder struct {
 	runner multistep.Runner
 }
 
+func (b *Builder) ConfigSpec() hcldec.ObjectSpec {
+	return b.config.FlatMapstructure().HCL2Spec()
+}
+
 // Prepare processes the build configuration parameters and validates the configuration
-func (self *Builder) Prepare(raws ...interface{}) (parms []string, err error) {
+func (b *Builder) Prepare(raws ...interface{}) ([]string, []string, error) {
+	var err error
 	// Parse and create the configuration
-	self.config, err = NewConfig(raws...)
+	b.config, err = NewConfig(raws...)
 
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// Check that the client/service is usable
-	service := self.config.GetService()
+	service := b.config.GetService()
 
 	if _, err := service.GetAccount(); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// Check that the specified storage device is a template
 	storageDetails, err := service.GetStorageDetails(&request.GetStorageDetailsRequest{
-		UUID: self.config.StorageUUID,
+		UUID: b.config.StorageUUID,
 	})
 
 	if err == nil && storageDetails.Type != upcloud.StorageTypeTemplate {
 		err = fmt.Errorf("The specified storage UUID is of invalid type \"%s\"", storageDetails.Type)
 	}
 
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
-	return nil, nil
+	return nil, nil, nil
 }
 
 // Run executes the actual build steps
-func (self *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
+func (b *Builder) Run(ctx context.Context, ui packer.Ui, hook packer.Hook) (packer.Artifact, error) {
```

Comment:
Same here, why change `self` to `b`?

Reply:
Just IDE recommendation

Reply:
"self", "this" is not the best practices for Golang receivers anyway...